File: #0085.maximal-rectangle.cpp
class Solution {
public:
    int maximalRectangle(vector<vector<char>>& matrix) {

        int rowSize = matrix.size();
        if (rowSize == 0) {
            return 0;
        }

        int colSize = matrix[0].size();
        vector<vector<int>> v(rowSize, vector<int>(colSize));
        for (int i = 0; i < rowSize; ++i) {
            for (int j = 0; j < colSize; ++j) {
                v[i][j] = (matrix[i][j] == '1') ? (j == 0 ? 1 : v[i][j - 1] + 1) : 0;
            }
        }

        int result = 0;
        int area;
        for (int i = 0; i < rowSize; ++i) {
            for (int j = 0; j < colSize; ++j) {
                area = INT_MAX;
                for (int k = i; k < rowSize; k++) {
                    area = std::min(area, v[k][j]);
                    if (area == 0) {
                        break;
                    }
                    result = std::max(area * (k - i + 1), result);
                }
            }
        }
        return result;
    }
};

File: #0085.maximal-rectangle.js
/**
 * @param {character[][]} matrix
 * @return {number}
 */
var maximalRectangle = function(matrix) {
    if (!matrix.length) {
        return 0;
    }

    const row = matrix.length;
    const col = matrix[0].length;
    const arr = Array(col).fill(0);
    let result = 0;
    for (let i = 0; i < row; i++) {
        for (let j = 0; j < col; j++) {
            arr[j] = +matrix[i][j] ? arr[j] + 1 : 0;
        }
        
        for(let j = 0; j < col; j++) {
            let left = j - 1;
            while (left >= 0 && arr[left] >= arr[j]) {
                left--;
            }

            let right = j + 1;
            while (right < col && arr[right] >= arr[j]) {
                right++;
            }
            result = Math.max(arr[j] * (right - left - 1), result);
        }
    }
    return result;
};
